### ISSUE TITLE
TestClock hangs with background thread

### ISSUE THREAD

**Issue by markusa380** [zio/zio#5178]
The example below is a simplification, in our real code the background worker does more important things than this of course, things that have a direct effect on the test succeeding.

```scala

import zio._
import zio.duration._
import zio.test.environment.TestClock
import zio.test.{DefaultRunnableSpec, ZSpec, assertCompletes}

import java.time.{OffsetDateTime, ZoneOffset}

object TestClockHangup extends DefaultRunnableSpec {

  val backgroundWorker = (
    for {
      _ <- ZIO(1 + 1).forever.fork
    } yield ()
  ).toLayer

  def spec: ZSpec[Environment, Failure] = suite("TestClockHangup")(
    testM("test clock hangup") {
      println("Before clock hangup")
      for {
        testClock <- ZIO.service[TestClock.Service]
        _         <- testClock.adjust(2.seconds)
        _         <- testClock.setDateTime(OffsetDateTime.of(2012, 12, 12, 12, 12, 12, 12, ZoneOffset.UTC))
        _          = println("This will never occur")
      } yield assertCompletes
    },
  ).provideCustomLayer(backgroundWorker)
}
```

We are now restructuring our tests to not require background workers anymore, but that's not the point.

The real problem is that `TestClock`, without explanation or good reason, waits that all threads in the test are either suspended or completed, instead of just the threads it has an effect on.

Maybe that is a technical limitation as one might not know which threads that would be, but the problem is that there's no information on what is happening.

It might make sense to introduce a warning that `TestClock` is waiting on a thread after a certain amount of time, and to maybe return a stack trace of the thread(s) in question, as my colleagues and I lost a full work day on this.

**Comment by adamgfraser** [comment]
@markusa380 The `TestClock` needs to wait for fibers to complete or enter a suspended state to avoid race conditions where the clock is advanced "too early" while effects that are relying on the prior state are still being executed. While theoretically it would be safe to advance the test clock if a fiber was still executing that had no effect on the test, I don't think it is possible to determine that when arbitrary effects can be performed within a `ZIO` workflow (e.g. the "background worker" could modify some arbitrary piece of state that influences the activity of one of the fibers in the test.

That being said, agree that a warning if the `TestClock` is waiting too long for the system to enter a suspended state would be useful.

**Comment by markusa380** [comment]
@adamgfraser Yes that's what we concluded in our research as well.

I'd say it's equally important to make it easy to find out _which_ threads are blocking the test clock, not just the fact that they exist.
Current solution we came up with (warning: ugly):

```
  def debugAllBlockingFibers(clock: TestClock.Service): UIO[Unit] = {
    def isRunningOrFinishing(fiber: Fiber.Status): Boolean = fiber match {
      case Running(_)   => true
      case Finishing(_) => true
      case _            => false
    }

    val testClock = clock.asInstanceOf[TestClock.Test]
    for {
      fibers <- testClock.supervisedFibers
      _      <- ZIO.foreach(fibers)(f =>
                  for {
                    trace  <- f.trace
                    status <- f.status
                  } yield
                    if (isRunningOrFinishing(status)) {
                      println(s"############# Thread is status $status: ##############\n\n${trace.prettyPrint}")
                    },
                )
    } yield ()
  }
```

**Comment by adamgfraser** [comment]
@markusa380 You can use `TestAspect.diagnose` for that.

**Comment by markusa380** [comment]
@adamgfraser Is there some surefire way to have background threads that are not taken into account in `TestClock`?
We tested `.forkDaemon` but it still takes the threads into account.

**Comment by adamgfraser** [comment]
@markusa380 No. The concept of a `Supervisor` is distinct from the concept of a `Scope` that a fiber is forked in. You can imagine that if the test clock didn't want for fibers forked with `forkDaemon` we would have another issue where someone's test wasn't working because of that. We are exploring for ZIO 2.0 a concept that would allow creating regions that are "shielded" from supervision but for now I would recommend avoiding using background fibers like this in tests where you are using the `TestClock`.

**Comment by markusa380** [comment]
@adamgfraser Thanks, that's important input for my team discussing how to adjust the tests.
